chia/cmds/param_types.py

```diff
@@ -0,0 +1,164 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from decimal import Decimal
+from typing import Any, Optional
+
+import click
+
+from chia.cmds.units import units
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.util.bech32m import bech32_decode, decode_puzzle_hash
+from chia.util.config import selected_network_address_prefix
+from chia.util.ints import uint64
+from chia.wallet.util.address_type import AddressType
+
+
+class TransactionFeeParamType(click.ParamType):
+    """
+    A Click parameter type for transaction fees, which can be specified in XCH or mojos.
+    """
+
+    name: str = "uint64"  # output type
+    value_limit: Decimal = Decimal(0.5)
+
+    def convert(self, value: Any, param: Optional[click.Parameter], ctx: Optional[click.Context]) -> uint64:
+        if isinstance(value, uint64):  # required by click
+            return value
+        mojos = False  # TODO: Add unit logic
+        if mojos:
+            try:
+                return uint64(int(value))
+            except ValueError:
+                self.fail("Fee must be integer value in mojos", param, ctx)
+        else:
+            try:
+                if self.value_limit != Decimal(0) and Decimal(value) > self.value_limit:
+                    self.fail(f"Fee must be in the range 0 to {self.value_limit}", param, ctx)
+                return uint64(int(Decimal(value) * units["chia"]))
+            except ValueError:
+                self.fail("Fee must be decimal dotted value in XCH (e.g. 0.00005)", param, ctx)
+
+
+@dataclass(frozen=True)
+class CliAmount:
+    """
+    A dataclass for TX / wallet amounts for both XCH and CAT, and of course mojos.
+    """
+
+    mojos: bool
```

Comment:
you already have this bool encoded in the `Optional` members below. except there are two of them and both being `None` or both being set violates the invariant